### PR TITLE
policy: make ToServices selectors work for in-cluster services too

### LINKED ISSUE #34021: Enhancement: make ToServices selectors work for in-cluster services too / add FromServices
CiliumNetworkPolicy supports selecting [peers by service](https://docs.cilium.io/en/latest/security/policy/language/#services-based):

```yaml
spec:
  endpointSelector:
    matchLabels: {}
  egress:
  - toServices:
    - k8sService:
        serviceName: myservice
        namespace: default
```

However, this has a significant limitation, in that it only supports selectorless services with Endpoints external to the cluster. In other words, the service acts as a sort of CIDRGroup. Internally, the policy is translated in to a ToCIDR selector with the IPs of the endpoints.

It would be nice if this limitation was relaxed. Instead, we could translate a label-selected service to its component endpointSelector. We already have most of the infrastructure for this -- the policy ingestion layer in `pkg/policy/k8s` already watches all services and already has most of the needed logic.

Additionally, toServices cannot be used with toPorts. This is an artificial limitation that should be removed.

-----

To generalize `toServices`, we need some code that:
1. Translates a `toServices` block in to *either* a set of CIDR selectors or an endpointSelector, depending on whether or not the service has a selector. The endpointSelector should be copied from the Service's selector
2. Tracks all policy to service mappings, and efficiently updates policies when services change, are created, or are deleted
3. Removes the restriction that `toServices` and `toPorts` cannot be used together.
4. (optional) adds a `fromServices` block that works the same as `toServices`

### PR DESCRIPTION
**Pull request by chaunceyjiang**
Fixes: #34021

```release-note
policy: make ToServices selectors work for in-cluster services too
```

### CODE REVIEW

File: pkg/policy/k8s/service.go
Comment (squeed) [inline pkg/policy/k8s/service.go]: I think this whole function can just be `return svc != nil` -- or even just removed.

Comment (chaunceyjiang) [reply pkg/policy/k8s/service.go]: Updated.

**Comment by squeed** [comment]
This needs a unit test :)

**Comment by chaunceyjiang** [comment]
@squeed PTAL.

**Review by learnitall — APPROVED** [review]
API changes LGTM, thanks. Do note though that my familiarity with this code is limited, I'd take my approval with a grain of salt.

File: pkg/policy/k8s/service.go
Comment (doniacld) [inline pkg/policy/k8s/service.go]: Is the comment still accurate?

**Review by doniacld — APPROVED** [review]
One minor comment otherwise LGTM!

File: pkg/policy/k8s/service.go
Comment (chaunceyjiang) [reply pkg/policy/k8s/service.go]: Thanks for the reminder. I’ll delete this line.

File: pkg/policy/k8s/service_test.go
Comment (odinuge) [inline pkg/policy/k8s/service_test.go]: I'm not 100% in sync with what the intention of this is.

Afaik. this will just make the label selection match all _pods_ (endpoints for actual pods) in all namespaces that match the pod selector in the service (this without any explicit namespace matching??)? I don't think that is the case we want here, or am I missing something?


I think this deserves another test as well with pod in namespace A, selecting a service in namespace B - to properly ensure that the namespaces are handled correctly.



So if I read this correctly, this will result in;
- K8s service X in namespace A selecting `foo=bar`
- Pod P in namespace A
- CNP in Namespace A selecting P, with toService to X
- This will open egress to all pods in the cluster with label `foo=bar`?

File: pkg/policy/k8s/service.go
Comment (odinuge) [inline pkg/policy/k8s/service.go]: How is the namespace of the k8s service matched here plumbed into the selector?

As I understand the code by looking briefly at it, the namespace of the service is not part of the `s.svc.Selector` - and it will therefore not be part of the `endpointSelector` here. Is that assumption false?

File: pkg/policy/k8s/service.go
Comment (odinuge) [inline pkg/policy/k8s/service.go]: Are we 100% sure this is safe? By reading this, my intuition says that this line "ignores" delete events. eg. - but I don't think that is the case. Looking a bit below, the existing line `newSVC.DeepEqual(oldSVC)` implicitly says that newSVC is always non-nil, because otherwise it will be a null pointer panic. I defer to other cilium maintainers how to deal with this, but it seems strange to have a silent`if newSVC == nil` if we never send in `nil`, and I think it in the future can hide resource leaks.

**Review by odinuge — COMMENTED** [review]
Some question about the intention here, and the thoughts around namespaces. Maybe I'm missing some context, but I think its useful to square them out.

I also think its worth mentioning how this is supposed to work with headless services where `clusterIP: None`, and add a test showing that the implementation works the same as the intention. There are a lot of components at play here, so its a bit tricky to test this properly without having an integration/e2e test imo. 


Thanks for working on this! I'm really excited about being able to use this!

File: pkg/policy/k8s/service.go
Comment (odinuge) [reply pkg/policy/k8s/service.go]: I'm also pretty worried about plumbing this directly without sanitizing it correctly. I think that can result in potentially undesired behavior with certain labels. It could have been sanitized earlier, but we have to be 100% sure that is the case, and that edge cases are properly handled, before adding it like this.


I guess one use-case here is;
- Admin creates CNP in namespace A to select all services in namespace
- User has access to create Kubernetes services in namespace A, but _not_ endpoints/endpointslices or network polices of any kind.

In this case I believe the user should only be able to open egress/ingress to/from pods in namespace A, and not other namespaces or other endpoints in the cluster.

cc @joestringer

File: pkg/policy/k8s/service_test.go
Comment (chaunceyjiang) [reply pkg/policy/k8s/service_test.go]: https://github.com/cilium/cilium/blob/b35017b01c0f9893bf4e6a74182f171091df2f07/pkg/k8s/apis/cilium.io/utils/utils.go#L83-L95

If I understand correctly, it should be in the same namespace.

**Comment by odinuge** [comment]
I think its also important to note in docs (and potentially changelog) that this requires the service selector to use only labels that are [identity-Relevant Labels](https://docs.cilium.io/en/stable/operations/performance/scalability/identity-relevant-labels/) in order to match the endpoints, and that its not sufficient that the IPs are added to the `endpoint` or `endpointslices` objects for the service.

File: pkg/policy/k8s/service.go
Comment (chaunceyjiang) [inline pkg/policy/k8s/service.go]: @odinuge PTAL.

> Afaik. this will just make the label selection match all pods (endpoints for actual pods) in all namespaces that match the pod selector in the service (this without any explicit namespace matching??)? I don't think that is the case we want here, or am I missing something?

I think you're right. we should add a selector for the namespace.

File: pkg/policy/k8s/service.go
Comment (odinuge) [inline pkg/policy/k8s/service.go]: It possible that the namespace is set to an empty string, and that will match services in all namespaces with the name given - but not all endpoints matching the service selector clusterwide.

So I think here you should use ~`s.svc.Namespace`~ `s.svcID.Namespace` instead - and we know a service is always namespaced, so that should never be an empty string.

Comment (chaunceyjiang) [reply pkg/policy/k8s/service.go]: Done.

File: pkg/policy/k8s/service.go
Comment (squeed) [reply pkg/policy/k8s/service.go]: Very good question; it seems like @chaunceyjiang has since changed the code to force-add a namespace selector, which mimics what the services controller would do.

So, yes, good catch!

**Comment by squeed** [comment]
> I think its also important to note in docs (and potentially changelog) that this requires the service selector to use only labels that are [identity-Relevant Labels](https://docs.cilium.io/en/stable/operations/performance/scalability/identity-relevant-labels/) in order to match the endpoints, and that its not sufficient that the IPs are added to the `endpoint` or `endpointslices` objects for the service.

Good observation. In theory, we could drop all non-identity-relevant labels from the selector, but we don't do that for hand-written label selectors. So, if we keep with the model of "toServices just copies selectors", then "fixing" it in documentation is more correct.

File: pkg/policy/k8s/service.go
Comment (squeed) [reply pkg/policy/k8s/service.go]: My reading of this is also true; we would not update dependent services on service delete. So yes, this needs to be fixed.

File: pkg/policy/k8s/service.go
Comment (squeed) [inline pkg/policy/k8s/service.go]: can `svc` ever be nil? I hope not!

File: pkg/policy/k8s/service.go
Comment (squeed) [inline pkg/policy/k8s/service.go]: nit: write an `appendSelector()` function for consistency :-)

File: pkg/policy/k8s/service_test.go
Comment (squeed) [inline pkg/policy/k8s/service_test.go:458]: Can you validate the following cases:

1. Add policy with non-matching service selector
2. Add service that matches, validate selector is correct
3. Update matched service, validate selector is correct
4. Add second matching service, validate selector
5. Delete each service, validating generated selector
6. Delete policy
7. Add service
8. Add policy, validate selection is correct

Make sense?

File: pkg/policy/k8s/service.go
Comment (chaunceyjiang) [reply pkg/policy/k8s/service.go]: So you're saying I don't need to check if it's nil, right?

Comment (squeed) [reply pkg/policy/k8s/service.go]: Worth a quick check, but I don’t think so. It comes from the ServiceCache.

File: pkg/policy/k8s/service_test.go
Comment (chaunceyjiang) [reply pkg/policy/k8s/service_test.go:458]: Done.

**Comment by chaunceyjiang** [comment]
@squeed PTAL.

**Comment by chaunceyjiang** [comment]
> Looks good! The documentation will need to be updated; would you like to do that in this PR or in a follow-up?

I'll update it in the next PR.
/cc @squeed

**Comment by chaunceyjiang** [comment]
Hi, @squeed  The ci-ginkgo test in the E2E pipeline is failing.
https://github.com/cilium/cilium/actions/runs/10958473948/job/30428853645

However, when I run the test manifest locally at https://github.com/cilium/cilium/blob/main/test/k8s/manifests/high-scale-ipcache.yaml, it works fine in my environment.

``` 
# kubectl get pods -l type=client                                       1 ↵
NAME                      READY   STATUS    RESTARTS      AGE
client-6f594c4976-pmgf7   1/1     Running   1 (31s ago)   34m
client-6f594c4976-rlmb5   1/1     Running   1 (32s ago)   34m
client-6f594c4976-tclsg   1/1     Running   1 (31s ago)   34m
client-6f594c4976-tfgjl   1/1     Running   1 (32s ago)   34m
```

https://github.com/cilium/cilium/blob/main/test/k8s/datapath_configuration.go#L686-L687
What's the purpose of using `bpftool map update`?

Can you take another look for me? Thanks~

**Comment by chaunceyjiang** [comment]
@squeed PTAL.

**Comment by squeed** [comment]
This looks good to me!

I would like to get https://github.com/cilium/cilium/pull/35293 merged first, as it needs to be backported and there will be a merge conflict. But then this can go in :-).

**Comment by squeed** [comment]
Getting there. The blocking PR is still hung on reviews, then this can go in :-).

**Review by squeed — APPROVED** [review]

**Comment by squeed** [comment]
@chaunceyjiang nice work :). Once this lands, we also need to update the docs.

**Comment by chaunceyjiang** [comment]
>  Once this lands, we also need to update the docs.

OK.

**Comment by chaunceyjiang** [comment]
@squeed Hi, is the `fromServices block` mentioned in the issue https://github.com/cilium/cilium/issues/34021 still needed to be pushed forward? If needed, I will complete it #34627 after finishing the documentation for this PR, .

**Comment by joestringer** [comment]
I've relabeled the release note category for this to "major" as I think this is a significant new change that users may be interested to find out about and start to use. I think it would be neat to feature this in the release announcement for the upcoming v1.17 release ETA early next year.

**Pull request merged**
